File: frontend/lib/toon.ts
/**
 * TOON - Token-Oriented Object Notation (TypeScript)
 * Mirrors backend/app/services/toon.py
 */

export function needsQuote(s: string): boolean {
  if (s === "") return true;
  if (s === "true" || s === "false" || s === "null") return true;
  if (/^-?\d+(\.\d+)?$/.test(s)) return true;
  if (s.includes("\n") || s.includes(":") || s.includes("#") || [" ", "-", '"', "'"].includes(s[0])) return true;
  return false;
}

export function encodeScalar(v: any): string {
  if (v === null || v === undefined) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return String(v);
  if (typeof v === "string") {
    if (needsQuote(v)) {
      const esc = v.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n");
      return `"${esc}"`;
    }
    return v;
  }
  return String(v);
}

export function decodeScalar(s: string): any {
  s = s.trim();
  if (s === "null") return null;
  if (s === "true") return true;
  if (s === "false") return false;
  if (/^-?\d+$/.test(s)) return parseInt(s, 10);
  if (/^-?\d+\.\d+$/.test(s)) return parseFloat(s);
  if (s.length >= 2 && s[0] === '"' && s[s.length - 1] === '"') {
    return s.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, "\\").replace(/\\n/g, "\n");
  }
  return s;
}

export function dumps(obj: any, indent = 0): string {
  const lines: string[] = [];
  dumpValue(obj, lines, indent, undefined, true);
  return lines.join("\n");
}

function dumpValue(obj: any, lines: string[], indent: number, key?: string, top?: boolean) {
  const prefix = "  ".repeat(indent);
  if (obj !== null && typeof obj === "object" && !Array.isArray(obj)) {
    const keys = Object.keys(obj);
    if (keys.length === 0) {
      if (key !== undefined) lines.push(`${prefix}${key}: {}`);
      return;
    }
    if (top) {
      for (const k of keys) dumpValue(obj[k], lines, indent, k, false);
    } else {
      if (key !== undefined) {
        lines.push(`${prefix}${key}:`);
        const base = indent + 1;
        for (const k of keys) dumpValue(obj[k], lines, base, k, false);
      } else {
        for (const k of keys) dumpValue(obj[k], lines, indent, k, false);
      }
    }
  } else if (Array.isArray(obj)) {
    if (key !== undefined) {
      if (obj.length === 0) {
        lines.push(`${prefix}${key}: []`);
        return;
      }
      lines.push(`${prefix}${key}:`);
      const base = indent + 1;
      for (const item of obj) {
        if (item !== null && typeof item === "object" && !Array.isArray(item)) {
          let first = true;
          for (const [dk, dv] of Object.entries(item)) {
            if (first) {
              if (dv !== null && typeof dv === "object") {
                lines.push(`${"  ".repeat(base)}- ${dk}:`);
                dumpValue(dv, lines, base + 1, undefined, false);
              } else {
                lines.push(`${"  ".repeat(base)}- ${dk}: ${encodeScalar(dv)}`);
              }
              first = false;
            } else {
              if (dv !== null && typeof dv === "object") {
                lines.push(`${"  ".repeat(base + 1)}${dk}:`);
                dumpValue(dv, lines, base + 2, undefined, false);
              } else {
                lines.push(`${"  ".repeat(base + 1)}${dk}: ${encodeScalar(dv)}`);
              }
            }
          }
          if (first) lines.push(`${"  ".repeat(base)}- {}`);
        } else if (Array.isArray(item)) {
          lines.push(`${"  ".repeat(base)}-`);
          dumpValue(item, lines, base + 1, undefined, false);
        } else {
          lines.push(`${"  ".repeat(base)}- ${encodeScalar(item)}`);
        }
      }
    } else {
      for (const item of obj) dumpValue(item, lines, indent, undefined, false);
    }
  } else {
    if (key !== undefined) lines.push(`${prefix}${key}: ${encodeScalar(obj)}`);
    else lines.push(`${prefix}${encodeScalar(obj)}`);
  }
}

export function loads(text: string): any {
  if (!text || !text.trim()) return {};
  const rawLines = text.split("\n");
  const cleaned: string[] = [];
  for (const l of rawLines) {
    if (!l.trim()) continue;
    const stripped = l.trimStart();
    if (stripped.startsWith("#")) continue;
    cleaned.push(l);
  }
  if (cleaned.length === 0) return {};
  const [result] = parseBlock(cleaned, 0, 0);
  return result;
}

function countIndent(line: string): number {
  let c = 0;
  for (const ch of line) if (ch === " ") c++; else break;
  return Math.floor(c / 2);
}

function splitKv(s: string): [string, string | null] {
  const idx = s.indexOf(":");
  if (idx === -1) return [s.trim(), null];
  const k = s.slice(0, idx).trim();
  const v = s.slice(idx + 1).trim();
  if (v === "") return [k, ""];
  return [k, v];
}

function parseBlock(lines: string[], start: number, baseIndent: number): [any, number] {
  let obj: any = {};
  let isArray = false;
  let arrayItems: any[] = [];
  let i = start;
  while (i < lines.length) {
    const line = lines[i];
    if (!line.trim()) { i++; continue; }
    const indent = countIndent(line);
    if (indent < baseIndent) break;
    if (indent > baseIndent) break;
    const stripped = line.trimStart();
    if (stripped.startsWith("- ")) {
      isArray = true;
      const content = stripped.slice(2);
      if (!content || content.trim() === "") {
        if (i + 1 < lines.length && countIndent(lines[i + 1]) > baseIndent) {
          const [sub, nxt] = parseBlock(lines, i + 1, baseIndent + 1);
          arrayItems.push(sub);
          i = nxt;
        } else {
          arrayItems.push(null);
          i++;
        }
      } else if (content.trim().startsWith('"') || content.trim().startsWith("'")) {
        // quoted scalar array item may contain colon — treat as string
        arrayItems.push(decodeScalar(content.trim()));
        i++;
      } else if (content.includes(":")) {
        const d: any = {};
        const [k, v] = splitKv(content);
        if (v === "" || v === null) {
          if (i + 1 < lines.length && countIndent(lines[i + 1]) > baseIndent) {
            const nextIndent = countIndent(lines[i + 1]);
            if (nextIndent === baseIndent + 1 && !lines[i + 1].trimStart().startsWith("- ")) {
              d[k] = v === "" ? null : decodeScalar(v!);
              let j = i + 1;
              while (j < lines.length) {
                if (countIndent(lines[j]) !== baseIndent + 1) break;
                if (lines[j].trimStart().startsWith("- ")) break;
                const ns = lines[j].trimStart();
                if (!ns.includes(":")) break;
                const [nk, nv] = splitKv(ns);
                if (nv === "" || nv === null) {
                  if (j + 1 < lines.length && countIndent(lines[j + 1]) > baseIndent + 1) {
                    const [sub2, nxt2] = parseBlock(lines, j + 1, baseIndent + 2);
                    d[nk] = sub2;
                    j = nxt2;
                  } else {
                    d[nk] = {};
                    j++;
                  }
                } else {
                  d[nk] = decodeScalar(nv);
                  j++;
                }
              }
              arrayItems.push(d);
              i = j;
              continue;
            } else {
              d[k] = {};
              arrayItems.push(d);
              i++;
            }
          } else {
            d[k] = {};
            arrayItems.push(d);
            i++;
          }
        } else {
          d[k] = decodeScalar(v);
          let j = i + 1;
          while (j < lines.length) {
            if (countIndent(lines[j]) !== baseIndent + 1) break;
            if (lines[j].trimStart().startsWith("- ")) break;
            const ns = lines[j].trimStart();
            if (!ns.includes(":")) break;
            const [nk, nv] = splitKv(ns);
            if (nv === "" || nv === null) {
              if (j + 1 < lines.length && countIndent(lines[j + 1]) > baseIndent + 1) {
                const [sub2, nxt2] = parseBlock(lines, j + 1, baseIndent + 2);
                d[nk] = sub2;
                j = nxt2;
              } else {
                d[nk] = {};
                j++;
              }
            } else {
              d[nk] = decodeScalar(nv);
              j++;
            }
          }
          arrayItems.push(d);
          i = j;
          continue;
        }
      } else {
        arrayItems.push(decodeScalar(content));
        i++;
      }
    } else {
      if (!stripped.includes(":")) { i++; continue; }
      const [k, v] = splitKv(stripped);
      const hasNested = i + 1 < lines.length && countIndent(lines[i + 1]) > baseIndent;
      if (v === "" || v === null) {
        if (hasNested) {
          const nextStripped = lines[i + 1].trimStart();
          if (nextStripped.startsWith("- ")) {
            const [sub, nxt] = parseBlock(lines, i + 1, baseIndent + 1);
            obj[k] = sub;
            i = nxt;
          } else {
            const [sub, nxt] = parseBlock(lines, i + 1, baseIndent + 1);
            obj[k] = sub;
            i = nxt;
          }
        } else {
          obj[k] = {};
          i++;
        }
      } else {
        if (v === "[]") { obj[k] = []; i++; }
        else if (v === "{}") { obj[k] = {}; i++; }
        else { obj[k] = decodeScalar(v); i++; }
      }
    }
  }
  if (isArray) return [arrayItems, i];
  return [obj, i];
}

export function validate(text: string): boolean {
  try { loads(text); return true; } catch { return false; }
}

export function estimateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

export function tokenSavings(jsonText: string, toonText: string) {
  const jt = estimateTokens(jsonText);
  const tt = estimateTokens(toonText);
  const saving = jt > 0 ? ((jt - tt) / jt) * 100 : 0;
  return { json_tokens: jt, toon_tokens: tt, saving_percent: Math.round(saving * 10) / 10 };
}

export function jsonToToon(jsonText: string): string {
  const obj = JSON.parse(jsonText);
  return dumps(obj);
}

export function toonToJson(toonText: string): string {
  const obj = loads(toonText);
  return JSON.stringify(obj, null, 2);
}

File: backend/app/services/toon.py
"""
TOON - Token-Oriented Object Notation
Compact, line-oriented alternative to JSON. ~30-60% fewer tokens.
Spec:
- scalars: string (unquoted if safe), number, boolean, null
- `key: value` for scalars
- `key:` for nested objects (indented 2 spaces)
- `key:` + indented `- item` for arrays
- arrays of objects: `- key: value`
- strings with special chars quoted with double quotes
"""
from __future__ import annotations
import re
from typing import Any

SAFE_KEY_RE = re.compile(r'^[a-zA-Z_][a-zA-Z0-9_\-\.]*$')
SAFE_VAL_RE = re.compile(r'^[a-zA-Z0-9_\-\.\/]+$')

def _needs_quote(s: str) -> bool:
    if s == "":
        return True
    if s in ("true", "false", "null"):
        return True
    if re.match(r'^-?\d+(\.\d+)?$', s):
        return True
    if "\n" in s or ":" in s or "#" in s or s[0] in (" ", "-", '"', "'"):
        return True
    return False

def _encode_scalar(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, str):
        if _needs_quote(v):
            # escape double quotes
            escaped = v.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
            return f'"{escaped}"'
        return v
    return str(v)

def _decode_scalar(s: str) -> Any:
    s = s.strip()
    if s == "null":
        return None
    if s == "true":
        return True
    if s == "false":
        return False
    if re.match(r'^-?\d+$', s):
        try:
            return int(s)
        except:
            pass
    if re.match(r'^-?\d+\.\d+$', s):
        try:
            return float(s)
        except:
            pass
    if len(s) >= 2 and s[0] == '"' and s[-1] == '"':
        inner = s[1:-1].replace('\\"', '"').replace("\\\\", "\\").replace("\\n", "\n")
        return inner
    return s

def dumps(obj: Any, indent: int = 0) -> str:
    """Serialize Python object to TOON string."""
    lines = []
    _dump_value(obj, lines, indent, top=True)
    return "\n".join(lines)

def _dump_value(obj: Any, lines: list[str], indent: int, key: str | None = None, top: bool = False):
    prefix = "  " * indent
    if isinstance(obj, dict):
        if not obj:
            if key is not None:
                lines.append(f"{prefix}{key}: {{}}")
            return
        if top:
            for k, v in obj.items():
                _dump_value(v, lines, indent, key=k, top=False)
        else:
            if key is not None:
                lines.append(f"{prefix}{key}:")
                base = indent + 1
            else:
                base = indent
            for k, v in obj.items():
                _dump_value(v, lines, base, key=k, top=False)
    elif isinstance(obj, list):
        if key is not None:
            if not obj:
                lines.append(f"{prefix}{key}: []")
                return
            lines.append(f"{prefix}{key}:")
            base = indent + 1
            for item in obj:
                if isinstance(item, dict):
                    # dict in array: "- key: value"
                    first = True
                    for dk, dv in item.items():
                        if first:
                            # first key on same line as dash
                            if isinstance(dv, (dict, list)):
                                lines.append(f"{'  ' * base}- {dk}:")
                                _dump_value(dv, lines, base + 1, key=None, top=False)
                            else:
                                lines.append(f"{'  ' * base}- {dk}: {_encode_scalar(dv)}")
                            first = False
                        else:
                            if isinstance(dv, (dict, list)):
                                lines.append(f"{'  ' * (base + 1)}{dk}:")
                                _dump_value(dv, lines, base + 2, key=None, top=False)
                            else:
                                lines.append(f"{'  ' * (base + 1)}{dk}: {_encode_scalar(dv)}")
                    if first:
                        lines.append(f"{'  ' * base}- {{}}")
                elif isinstance(item, list):
                    lines.append(f"{'  ' * base}-")
                    _dump_value(item, lines, base + 1, key=None, top=False)
                else:
                    lines.append(f"{'  ' * base}- {_encode_scalar(item)}")
        else:
            for item in obj:
                # Nested arrays still require an explicit dash. Omitting it
                # flattened values such as event.files during round-tripping.
                if isinstance(item, dict):
                    first = True
                    for dk, dv in item.items():
                        marker = "- " if first else "  "
                        item_indent = indent if first else indent + 1
                        if isinstance(dv, (dict, list)):
                            lines.append(f"{'  ' * item_indent}{marker}{dk}:")
                            _dump_value(dv, lines, indent + 1, key=None, top=False)
                        else:
                            lines.append(f"{'  ' * item_indent}{marker}{dk}: {_encode_scalar(dv)}")
                        first = False
                    if first:
                        lines.append(f"{'  ' * indent}- {{}}")
                elif isinstance(item, list):
                    lines.append(f"{'  ' * indent}-")
                    _dump_value(item, lines, indent + 1, key=None, top=False)
                else:
                    lines.append(f"{'  ' * indent}- {_encode_scalar(item)}")
    else:
        if key is not None:
            lines.append(f"{prefix}{key}: {_encode_scalar(obj)}")
        else:
            lines.append(f"{prefix}{_encode_scalar(obj)}")

def loads(text: str) -> Any:
    """Parse TOON string to Python object."""
    if not text or not text.strip():
        return {}
    lines = text.splitlines()
    # preprocess: remove empty lines and comments (lines starting with #)
    cleaned = []
    for l in lines:
        if not l.strip():
            continue
        stripped = l.lstrip()
        if stripped.startswith("#"):
            continue
        cleaned.append(l)
    if not cleaned:
        return {}
    result, _ = _parse_block(cleaned, 0, 0)
    return result

def _count_indent(line: str) -> int:
    c = 0
    for ch in line:
        if ch == " ":
            c += 1
        else:
            break
    return c // 2

def _parse_block(lines: list[str], start: int, base_indent: int):
    obj = {}
    is_array = False
    array_items: list[Any] = []
    i = start
    while i < len(lines):
        line = lines[i]
        if not line.strip():
            i += 1
            continue
        indent = _count_indent(line)
        if indent < base_indent:
            break
        if indent > base_indent:
            # this should be handled by parent caller that peeked ahead
            break
        stripped = line.lstrip()
        if stripped.startswith("- "):
            # array at current level
            is_array = True
            # ensure obj is empty if we switch to array
            content = stripped[2:]
            if not content or content.strip() == "":
                # nested block follows
                if i + 1 < len(lines) and _count_indent(lines[i+1]) > base_indent:
                    sub, nxt = _parse_block(lines, i+1, base_indent+1)
                    array_items.append(sub)
                    i = nxt
                else:
                    array_items.append(None)
                    i += 1
            elif content.strip().startswith('"') or content.strip().startswith("'"):
                # quoted scalar array item (may contain colon) — treat as string, not dict
                array_items.append(_decode_scalar(content.strip()))
                i += 1
            elif ":" in content:
                # could be dict item: "- key: value" or "- key:"
                # parse as dict with first key
                # create a dict for this item
                d: dict[str, Any] = {}
                # parse first kv
                k, v = _split_kv(content)
                if v == "" or v is None:
                    # value is nested block
                    if i + 1 < len(lines) and _count_indent(lines[i+1]) > base_indent:
                        # need to look ahead: is next line indented more than base?
                        # For dict in array, subsequent keys are at indent+1
                        # So parse remaining keys
                        # First, handle nested if value expects block
                        # Check if next line is indented base+1 and not array
                        nxt_indent = _count_indent(lines[i+1])
                        if nxt_indent == base_indent + 1 and not lines[i+1].lstrip().startswith("- "):
                            # next line is continuation of this dict? Actually keys at base+1 would be continuation
                            # But if v is empty, next block at base+1 belongs to this key
                            # Let's try to parse block for this key's value
                            # For simplicity, if v is empty and next is block, parse it
                            sub, nxt = _parse_block(lines, i+1, base_indent+1)
                            # But we need to differentiate: is sub a dict for the key's value or remaining keys?
                            # Heuristic: if sub is dict and we have no value, treat as value
                            # However for array-of-dicts like "- id: 102\n  created_at: 2026-08-20"
                            # The second line "  created_at: ..." is at same indent as first key's value? No.
                            # Actually "- id: 102" is at base, then "  created_at: 2026-08-20" is at base+1
                            # So after first key, we should collect additional keys at base+1
                            # Let's collect them
                            d[k] = _decode_scalar(v) if v not in ("", None) else None
                            # now collect subsequent keys at indent base+1 that are not dash
                            j = i + 1
                            while j < len(lines):
                                if _count_indent(lines[j]) != base_indent + 1:
                                    break
                                if lines[j].lstrip().startswith("- "):
                                    break
                                nxt_stripped = lines[j].lstrip()
                                if ":" not in nxt_stripped:
                                    break
                                nk, nv = _split_kv(nxt_stripped)
                                if nv == "" or nv is None:
                                    # nested
                                    if j + 1 < len(lines) and _count_indent(lines[j+1]) > base_indent + 1:
                                        sub2, nxt2 = _parse_block(lines, j+1, base_indent+2)
                                        d[nk] = sub2
                                        j = nxt2
                                    else:
                                        d[nk] = {}
                                        j += 1
                                else:
                                    # check for nested block after scalar? e.g., key: then next line list
                                    if j + 1 < len(lines) and ":" not in nv and _count_indent(lines[j+1]) == base_indent + 2 and lines[j+1].lstrip().startswith("-"):
                                        # array following
                                        sub2, nxt2 = _parse_block(lines, j+1, base_indent+2)
                                        # nv is empty string marker for array?
                                        # but nv is scalar already, so not
                                        d[nk] = _decode_scalar(nv)
                                        j += 1
                                    else:
                                        d[nk] = _decode_scalar(nv)
                                        j += 1
                            array_items.append(d)
                            i = j
                            continue
                        else:
                            d[k] = {}
                            array_items.append(d)
                            i += 1
                    else:
                        d[k] = {}
                        array_items.append(d)
                        i += 1
                else:
                    d[k] = _decode_scalar(v)
                    # collect subsequent keys at base+1
                    j = i + 1
                    while j < len(lines):
                        if _count_indent(lines[j]) != base_indent + 1:
                            break
                        if lines[j].lstrip().startswith("- "):
                            break
                        nxt_stripped = lines[j].lstrip()
                        if ":" not in nxt_stripped:
                            break
                        nk, nv = _split_kv(nxt_stripped)
                        if nv == "" or nv is None:
                            if j + 1 < len(lines) and _count_indent(lines[j+1]) > base_indent + 1:
                                sub2, nxt2 = _parse_block(lines, j+1, base_indent+2)
                                d[nk] = sub2
                                j = nxt2
                            else:
                                d[nk] = {}
                                j += 1
                        else:
                            d[nk] = _decode_scalar(nv)
                            j += 1
                    array_items.append(d)
                    i = j
                    continue
            else:
                # "- scalar"
                array_items.append(_decode_scalar(content))
                i += 1
        else:
            # "key: value" or "key:"
            if ":" not in stripped:
                # bare scalar at top level? treat as key with no value
                i += 1
                continue
            k, v = _split_kv(stripped)
            # peek next line
            has_nested = (i + 1 < len(lines) and _count_indent(lines[i+1]) > base_indent)
            if v == "" or v is None:
                if has_nested:
                    # check if nested is array (next line starts with -)
                    next_stripped = lines[i+1].lstrip()
                    if next_stripped.startswith("- "):
                        sub, nxt = _parse_block(lines, i+1, base_indent+1)
                        # sub could be array
                        if isinstance(sub, list):
                            obj[k] = sub
                        else:
                            # if array parser returned dict, convert? Should be list
                            obj[k] = sub if isinstance(sub, list) else array_items
                            # fallback
                            if isinstance(sub, dict) and array_items:
                                obj[k] = sub
                        i = nxt
                    else:
                        sub, nxt = _parse_block(lines, i+1, base_indent+1)
                        obj[k] = sub
                        i = nxt
                else:
                    # empty dict or list notation
                    obj[k] = {}
                    i += 1
            else:
                # v may be "[]" or "{}" or scalar, or inline array?
                if v == "[]":
                    obj[k] = []
                    i += 1
                elif v == "{}":
                    obj[k] = {}
                    i += 1
                elif v == "":
                    # treat as nested
                    if has_nested:
                        sub, nxt = _parse_block(lines, i+1, base_indent+1)
                        obj[k] = sub
                        i = nxt
                    else:
                        obj[k] = ""
                        i += 1
                else:
                    # scalar value, but could have nested block after? e.g., key: scalar with children? Not typical.
                    # check for special case where value is scalar but next lines are at indent+1 and belong to this key's dict? No.
                    obj[k] = _decode_scalar(v)
                    i += 1
    if is_array:
        return array_items, i
    return obj, i

def _split_kv(s: str) -> tuple[str, str | None]:
    idx = s.find(":")
    if idx == -1:
        return s.strip(), None
    k = s[:idx].strip()
    v = s[idx+1:].strip()
    # keep empty string as indicator for nested
    if v == "":
        return k, ""
    return k, v

def validate(text: str) -> bool:
    try:
        loads(text)
        return True
    except Exception:
        return False

def to_json_compatible(obj: Any) -> Any:
    return obj

def from_json(obj: Any) -> str:
    return dumps(obj)

def to_json(obj: Any) -> Any:
    # for Supabase storage: just return dict
    return obj

# Token count estimate (approx)
def estimate_tokens(text: str) -> int:
    # rough: 1 token ~ 4 chars
    return max(1, len(text) // 4)

def token_savings(json_text: str, toon_text: str) -> dict:
    jt = estimate_tokens(json_text)
    tt = estimate_tokens(toon_text)
    saving = (jt - tt) / jt * 100 if jt > 0 else 0
    return {"json_tokens": jt, "toon_tokens": tt, "saving_percent": round(saving, 1)}

# Convenience: json <-> toon conversion using json module
import json as _json

def json_to_toon(json_text: str) -> str:
    obj = _json.loads(json_text)
    return dumps(obj)

def toon_to_json(toon_text: str) -> str:
    obj = loads(toon_text)
    return _json.dumps(obj, indent=2)
